modules/standard/GPU.chpl

```diff
@@ -895,6 +895,55 @@ module GPU
 
   private import Time;
 
+  // We no doc it so we can test this independently
+  @chpldoc.nodoc
+  proc gpuCubSort(ref gpuInputArr : [] ?t) {
+    param cTypeName = if      t==int(8)   then "int8_t"
+                      else if t==int(16)  then "int16_t"
+                      else if t==int(32)  then "int32_t"
+                      else if t==int(64)  then "int64_t"
+                      else if t==uint(8)  then "uint8_t"
+                      else if t==uint(16) then "uint16_t"
+                      else if t==uint(32) then "uint32_t"
+                      else if t==uint(64) then "uint64_t"
+                      else if t==real(32) then "float"
+                      else if t==real(64) then "double"
+                      else                     "unknown";
+
+    if cTypeName == "unknown" {
+      compilerError("Arrays with ", t:string,
+                    " elements cannot be sorted with gpuCubSort functions");
+    }
+
+    // Only useful when calling gpuCubSort directly
+    extern proc chpl_gpu_can_cub_sort(): bool;
+    if !chpl_gpu_can_cub_sort() {
+      gpuSort(gpuInputArr);
+      return;
+    }
+
+    proc getExternFuncName(param op: string, type t) param: string {
+      return "chpl_gpu_sort_"+op+"_"+cTypeName;
+    }
+
+    // find the extern function we'll use
+    // (there's only one right now, the infrastructure allows more)
+    param externFunc = getExternFuncName("keys", t);
+    extern externFunc proc sort_fn(data, temp, size);
+
+    // Make another array which is needed for the sort
+    var temp : gpuInputArr.type;
+
+    const basePtr = c_ptrToConst(gpuInputArr);
+    const tempPtr = c_ptrTo(temp);
+    // Do the sort
+    sort_fn(basePtr, tempPtr, gpuInputArr.size);
+
+    // The sorted values are in temp, so we need to copy them back
+    // to the original array
+    gpuInputArr = temp;
```

Comment:
Hmm... can we do a `<=>` here? As much as it seems to be a redundant operation, what you have here will be a memcpy of the full data, whereas since `gpuInputArr` and `temp` are of same type and DefaultRectangular, I expect `<=>` to be just a pointer swap of implementation classes or the data buffer or something.

But premature optimization is root of all evil, so, maybe do that only if it helps with performance? And if it doesn't (or if you're short on time), maybe just add a comment here hinting at a potential optimization down the road?

Reply:
I'll add a comment for now just in case it causes issues in testing which I won't have time to resolve